dataplane/saiserver/l2.go

```diff
@@ -16,48 +16,198 @@ package saiserver
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
 
 	"github.com/openconfig/lemming/dataplane/saiserver/attrmgr"
+	fwdpb "github.com/openconfig/lemming/proto/forwarding"
 
 	saipb "github.com/openconfig/lemming/dataplane/proto/sai"
 )
 
+type l2mcGroupMember struct {
+	oid      uint64
+	groupId  uint64
+	outputId uint64
+}
+
 type l2mcGroup struct {
 	saipb.UnimplementedL2McGroupServer
 	mgr       *attrmgr.AttrMgr
 	dataplane switchDataplaneAPI
+	groups    map[uint64]map[uint64]*l2mcGroupMember // map[group id][member oid]
 }
 
 func newL2mcGroup(mgr *attrmgr.AttrMgr, dataplane switchDataplaneAPI, s *grpc.Server) *l2mcGroup {
 	mg := &l2mcGroup{
 		mgr:       mgr,
 		dataplane: dataplane,
+		groups:    map[uint64]map[uint64]*l2mcGroupMember{},
 	}
 	saipb.RegisterL2McGroupServer(s, mg)
 	return mg
 }
 
-// TODO: Implement this.
-func (mg *l2mcGroup) CreateL2McGroup(context.Context, *saipb.CreateL2McGroupRequest) (*saipb.CreateL2McGroupResponse, error) {
+func (mg *l2mcGroup) CreateL2McGroup(ctx context.Context, req *saipb.CreateL2McGroupRequest) (*saipb.CreateL2McGroupResponse, error) {
 	id := mg.mgr.NextID()
+	// Update internal data
+	mg.groups[id] = map[uint64]*l2mcGroupMember{}
+	// Update attributes
+	l2mcAttrs := &saipb.L2McGroupAttribute{
+		L2McOutputCount: proto.Uint32(0),
+		L2McMemberList:  []uint64{},
+	}
+	mg.mgr.StoreAttributes(id, l2mcAttrs)
 	return &saipb.CreateL2McGroupResponse{Oid: id}, nil
 }
 
-// TODO: Implement this.
-func (mg *l2mcGroup) CreateL2McGroupMember(context.Context, *saipb.CreateL2McGroupMemberRequest) (*saipb.CreateL2McGroupMemberResponse, error) {
+func (mg *l2mcGroup) portNidFromBrirdgeId(ctx context.Context, outputId uint64) (uint64, error) {
+	req := &saipb.GetBridgePortAttributeRequest{Oid: outputId, AttrType: []saipb.BridgePortAttr{saipb.BridgePortAttr_BRIDGE_PORT_ATTR_PORT_ID}}
+	resp := &saipb.GetBridgePortAttributeResponse{}
+	if err := mg.mgr.PopulateAttributes(req, resp); err != nil {
+		return 0, fmt.Errorf("failed to populate OutputId (oid=%d): %v", outputId, err)
+	}
+	if resp.GetAttr().PortId == nil {
+		return 0, fmt.Errorf("cannot find portId for bridge port %q", outputId)
+	}
+	return resp.GetAttr().GetPortId(), nil
+}
+
+// updateGroupMember updates the member of a L2MC group.
+// If m is nil, remove mid from the group(key: group id), otherwise add m to groups with mid as the key.
+func (mg *l2mcGroup) updateGroupMember(ctx context.Context, gid, mid uint64, m *l2mcGroupMember) error {
+	if m == nil {
+		// Remove the member.
+		delete(mg.groups[gid], mid)
+		if len(mg.groups[gid]) == 0 {
+			delete(mg.groups, gid)
+		}
+	} else {
+		// Add a member.
+		mg.groups[gid][mid] = m
+		attr := &saipb.L2McGroupMemberAttribute{
+			L2McGroupId:  &m.groupId,
+			L2McOutputId: &m.outputId,
+		}
+		mg.mgr.StoreAttributes(mid, attr)
+	}
+	mList := []uint64{}
+	for _, m := range mg.groups[gid] {
+		mList = append(mList, m.oid)
+	}
+	gAttr := &saipb.L2McGroupAttribute{}
```

Comment:
this is unnecessary, StoreAttribute only stores non-nil attributes

Reply:
fixed